OSSM-8507 Migration guide for multitenant 2.6 (#86917)
openshift/openshift-docs: modules/ossm-migrating-multitenant-with-cert-manager.adoc

// Module included in the following assemblies:
//
// * service-mesh-docs-main/multitenant/ossm-migrating-multitenant-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="migrating-multitenant-with-cert-manager_{context}""]
= Migrating a multitenant deployment with cert-manager

.Prerequisites

> * You have deployed a cluster on {ocp-product-title} 4.14 or later.

Reviewer: ```suggestion
* You have deployed {ocp-product-title} 4.14 or later.
```

+
.`ServiceMeshControlPlane` cert-manager configuration
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  ...
  security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.istio-system.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
    manageNetworkPolicy: false
----

.Procedure

The `bookinfo` example application is being used for demonstration purposes with a minimal example for the `Istio` resource. For more information on configuration differences between the {SMProduct} 2 `ServiceMeshControlPlane` resource and the {SMProduct} 3 `Istio` resource, see "ServiceMeshControlPlane resource to Istio resource fields mapping".

You can follow these same steps with your own workloads.

. Update the `istio-csr` deployment to include your {SMProduct} 3 control plane by running the following command:
+
[source,terminal]
----
  helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
      --install \
      --reuse-values \
      --namespace istio-system \
      --wait \

>       -set "app.istio.revisions={basic,istio-tenant-a}" <1>

Reviewer 2: ```suggestion
      --set "app.istio.revisions={basic,istio-tenant-a}" <1>
```

<2> controlplanes by default watch the entire cluster and when managing multiple controlplanes on a single cluster, you must narrow the scope of each controlplane by setting `discoverySelectors`. In this example, we use the label `tenant` but you can use any label or combination of labels that you choose.
<3> If you are migrating metrics and tracing, uncomment and update the `extensionProviders` fields according to your tracing and metrics configurations.

. Add your `tenant` label to each one of your dataplane namespaces by running the following command for each dataplane namespace:
+
[source,terminal]
----
$ oc label ns bookinfo tenant=tenant-a
----
+
[NOTE]
====
In {SMProduct} 2.6, namespaces were enrolled into the mesh by adding them to the `Service MeshMemberRoll` resource. In {SMProduct} 3, you must label each one of your dataplane namespaces with the `discoverySelectors.matchLabels` label.
====

== Migrating workloads

Now you can migrate your workloads from the {SMProduct} 2.6 controlplane to the {SMproduct} 3.0 controlplane.

[NOTE]
====

> You can migrate workloads and gateways separately.

Reviewer 2: Same comment here about mentioning that you can also migrate them "in any order"

   reviews-v1-686cd989df-9x59z.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v2-785b8b48fc-l7xkj.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v3-67889ffd49-7bhxn.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
----

.. Check which proxies have been migrated to the new {SMProduct} 3.0 control plane by running the following command:
+
[source,terminal]
----
$ istioctl ps --istioNamespace istio-system-tenant-a --revision istio-tenant-a
----
+
.Example output
+
[source, terminal]
----
NAME                                         CLUSTER        CDS        LDS        EDS        RDS        ECDS     ISTIOD                      VERSION
productpage-v1-7745c5cc94-wpvth.bookinfo     Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-5bbf98dccf-n8566     1.24.3
----

> . Verify the `bookinfo` application is still working correctly by running the following command:

Reviewer 2: Trying to make it more obvious to readers that they will need to verify their own applications are working correctly and the exact commands they will run to do that will be different. The following command only applies to the bookinfo example.
```suggestion
. Verify your application is still working correctly. For the `bookinfo` application, run the following command:
```

// Module included in the following assemblies:
//
// * service-mesh-docs-main/multitenant/ossm-migrating-multitenant-assembly.adoc

:_mod-docs-content-type: PROCEDURE
[id="migrating-multitenant-with-cert-manager_{context}""]
= Migrating a multitenant deployment with cert-manager

.Prerequisites

* You have deployed a cluster on {ocp-product-title} 4.14 or later.
* You are logged in to the {ocp-product-title} web console as a user with the cluster-admin role.
* You have completed the premigration checklists.
* You have the {SMProduct} {SMv2Version} Operator installed.
* You have the {SMProduct} 3 Operator installed.
* You have `IstioCNI` installed.
* You have `istioctl` installed.

> * You are using the cert-manager tool in a multitenant deployment.

Reviewer 2: ```suggestion
* You are using the cert-manager and istio-csr tools in a multitenant deployment.
```

    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.istio-system.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
    manageNetworkPolicy: false
----

.Procedure

The `bookinfo` example application is being used for demonstration purposes. You can follow these same steps with your own workloads.

. Update the `istio-csr` deployment to include your {SMProduct} 3 control plane by running the following command:
+
[source,terminal]
----
  helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
      --install \
      --reuse-values \
      --namespace istio-system \
      --wait \
      --set "app.istio.revisions={basic,ossm3-v1-24-1}" <1>
----
+
<1> The `app.istio.revisions` field needs to include your {SMProduct} 3.0 control plane revision _before_ you create your `Istio` resource so that proxies can properly communicate with the {SMProduct} 3.0 control plane.

. Create your `Istio` resource.
+
.Example `Istio` resource with cert-manager
+
[source,yaml]
----
apiVersion: sailoperator.io/v1alpha1
kind: Istio
metadata:
  name: istio-tenant-a
spec:
  namespace: istio-system-tenant-a <1>

>   version: v1.24.3

Author: There is. I don't know that I'll have time to set that up for GA, but I can shortly after GA.

        - matchLabels:
            tenant: tenant-a <2>
    #  extensionProviders:  <3>
    #    - name: prometheus
    #      prometheus: {}
    #    - name: otel
    #      opentelemetry:
    #       port: 4317
    #      service: otel-collector.opentelemetrycollector-3.svc.cluster.local
      global:
         caAddress: cert-manager-istio-csr.istio-system.svc:443
      pilot:
        env:
          ENABLE_CA_SERVER: "false"
----
+
<1> The `spec.namespace` field in your `Istio` resource must be the **same** namespace as your `ServiceMeshControlPlane`. If you set the `spec.namespace` field in your `Istio` resource to a different namespace than your `ServiceMeshControlPlane`, the migration will not work properly.
<2> controlplanes by default watch the entire cluster and when managing multiple controlplanes on a single cluster, you must narrow the scope of each controlplane by setting `discoverySelectors`. In this example, we use the label `tenant` but you can use any label or combination of labels that you choose.
<3> If you are migrating metrics and tracing, uncomment and update the `extensionProviders` fields according to your tracing and metrics configurations.

> . Add your `tenant` label to each one of your dataplane namespaces by running the following command:

Reviewer 2: This reads a bit like you just run the command once but you actually need to run it once for every dataplane namespace.
```suggestion
. Add your `tenant` label to each one of your dataplane namespaces by running the following command for each dataplane namespace:
```

+
[source,terminal]
----
$ oc label ns bookinfo tenant=tenant-a
----
+
[NOTE]
====
In {SMProduct} 2.6, namespaces were enrolled into the mesh by adding them to the `Service MeshMemberRoll` resource. In {SMProduct} 3, you must label each one of your dataplane namespaces with the `discoverySelectors.matchLabels` label.
====

== Migrating workloads

Now you can migrate your workloads from the {SMProduct} 2.6 controlplane to the {SMproduct} 3.0 controlplane.

[NOTE]
====
You can migrate workloads and gateways separately.
====

. Find the current `IstioRevision` for your {SMProduct} 3.0 controlplane by running the following command:
+
[source,terminal]
----
$ oc get istios istio-tenant-a
NAME             REVISIONS   READY   IN USE   ACTIVE REVISION   STATUS    VERSION   AGE

> istio-tenant-a   1           1       0        istio-tenant-a    Healthy   v1.24.1   30s

Reviewer 2: ```suggestion
istio-tenant-a   1           1       0        istio-tenant-a    Healthy   v1.24.3   30s
```

----
+
[NOTE]
====
The naming format of your revisions depends on which upgrade strategy you choose for your `Istio` instance.
====

. Copy the `ACTIVE REVISION` to use as your `istio.io/rev` label in the next step.

. Update injection labels on the `dataplane` namespace by running the following command:
+
[source,terminal]
----
$ oc label ns bookinfo istio.io/rev=istio-tenant-a maistra.io/ignore-namespace="true" --overwrite=true
----
+
This adds the following labels to the namespace:
+
* `istio.io/rev: istio-tenant-a` label: Ensures that any new pods that get created in that namespace connect to the {SMProduct} 3.0 proxy.
+

> * `maistra.io/ignore-namespace: "true"` label: Disables sidecar injection for {SMProduct} 2.6 proxies in the namespace so the {SMProduct} 2.6 stops injecting proxies in this namespace, and any new proxies are injected by the {SMProduct} 3.0 `controlplane`. Without this label, the {SMProduct} 2.6 injection webhook tries to inject the pod and it will connect to the {SMProduct} 2.6 proxy as well as refuse to start since it will have the {SMProduct} 2.6 Carrier Infrastructure Network (CNI) annotation.

Reviewer 2: ```suggestion
* `maistra.io/ignore-namespace: "true"` label: Disables sidecar injection for {SMProduct} 2.6 proxies in the namespace so {SMProduct} 2.6 stops injecting proxies in this namespace, and any new proxies are injected by {SMProduct} 3.0. Without this label, the {SMProduct} 2.6 injection webhook tries to inject the pod and the injected sidecar proxy will refuse to start since it will have both the {SMProduct} 2.6 and the {SMProduct} 3.0 Container Network Interface(CNI) annotations.
```

+
[NOTE]
====
Once you apply the `maistra.io/ignore-namespace` label, any new pod that gets created in the namespace will connect to the {SMProduct} 3.0 proxy. Workloads can still communicate with each other regardless of which control plane they are connected to.
====

> . Use the `curl` command for the `productpage` pod in the `bookinfo` application so proxies can still communicate with one another by running the following command:
> +
> [source,terminal]
> ----
> $ oc exec -it -n bookinfo deployments/productpage-v1 -c istio-proxy -- curl localhost:9080/productpage
> ----

Reviewer 2: I don't think this step is necessary
```suggestion
```

. Restart the workloads:
+

> .. You can restart workloads all at once so that the new pod will be injected with the {SMProduct} 3.0 proxy by running the follow command:

Reviewer 2: This command to restart all workloads only works for namespaces that only have deployments in them. It's good for bookinfo but we should note that.
```suggestion
.. You can restart all the workloads at once so that the new pods will be injected with the {SMProduct} 3.0 proxy. For this example with bookinfo, run the following command:
```

+
[source,terminal]
----
$ oc rollout restart deployments -n bookinfo
----

> .. Alternatively, you can restart each workload individually by running the following command for each workload:

Reviewer 2: Same comment here that the general advice is applicable to all workloads but the specific command is only applicable for this example
```suggestion
.. Alternatively, you can restart each workload individually. For this example with bookinfo, run the following command for each workload:
```

+
[source,terminal]
----
$ oc rollout restart deployments productpage-v1 -n bookinfo
----

. Wait for the `productpage` application to restart by running the following command:
+
[source,terminal]
----
$ oc rollout status deployment productpage-v1 -n bookinfo
----

== Validate Workload Migration

> . CHeck that the productpage app is connected to the new controlplane.

Reviewer 2: ```suggestion
. Check that your workload is connected to the new controlplane.
```

.. See which proxies are still connected to the 2.6 controlplane with `istioctl` by running the following command:
+
[source,terminal]
----
$ istioctl ps --istioNamespace istio-system-tenant-a --revision basic
----
+

> Here `basic` should be the name of your `ServiceMeshControlPlane`:

Reviewer 2: ```suggestion
Here `basic` should be the name of your `ServiceMeshControlPlane`.
```

   ratings-v1-d6f449f59-9rds2.bookinfo               Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v1-686cd989df-9x59z.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v2-785b8b48fc-l7xkj.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v3-67889ffd49-7bhxn.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
----

.. Check which proxies have been migrated to the new {SMProduct} 3.0 control plane by running the following command:
+
[source,terminal]
----
$ istioctl ps --istioNamespace istio-system-tenant-a --revision istio-tenant-a
----
+
.Example output
+
[source, terminal]
----
NAME                                         CLUSTER        CDS        LDS        EDS        RDS        ECDS     ISTIOD                      VERSION

> productpage-v1-7745c5cc94-wpvth.bookinfo     Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-5bbf98dccf-n8566     1.23.0

Reviewer 2: A version variable for service mesh 3 would help here too
```suggestion
productpage-v1-7745c5cc94-wpvth.bookinfo     Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-5bbf98dccf-n8566     1.24.3
```

* You created an `IstioCNI` resource.
* You have the `istioctl` tool installed.
* You are using the cert-manager and istio-csr tools in a multitenant deployment.
//change to "You are using the cert-manager and istio-csr tools in a cluster-wide deployment" for the cluster-wide procedure
* Your {SMProduct} 2 `ServiceMeshControlPlane` is configured with the cert-manager tool:
+
.`ServiceMeshControlPlane` cert-manager configuration
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  ...
  security:
    certificateAuthority:
      cert-manager:
        address: cert-manager-istio-csr.istio-system.svc:443
      type: cert-manager
    dataPlane:
      mtls: true
    identity:
      type: ThirdParty
    manageNetworkPolicy: false
----

.Procedure

. Update the `istio-csr` deployment to include your {SMProduct} 3 control plane by running the following command:
+
[source,terminal]
----
  helm upgrade cert-manager-istio-csr jetstack/cert-manager-istio-csr \
      --install \
      --reuse-values \
      --namespace istio-system \
      --wait \
      --set "app.istio.revisions={basic,istio-tenant-a}" <1>
----
+

> <1> The `app.istio.revisions` field needs to include your {SMProduct} 3.0 control plane revision _before_ you create your `Istio` resource so that proxies can properly communicate with the {SMProduct} 3.0 control plane.

Author: If they don't include the 3.0 control plane before they create their Istio resource, then they have to blow up their Istio resource and start over. This has happened to customers before, so emphasis has been added so it is clear this must be done first.

====

=== Migrating workloads

Now you can migrate your workloads from the {SMProduct} 2.6 control plane to the {SMproduct} 3.0 control plane.

[NOTE]
====
You can migrate workloads and gateways separately, and in any order. See "Migrating gateways".
====

. Find the current `IstioRevision` for your {SMProduct} 3.0 control plane by running the following command:
+
[source,terminal]
----
$ oc get istios istio-tenant-a
----
+
.Example output
+
[source,terminal]
----
NAME             REVISIONS   READY   IN USE   ACTIVE REVISION   STATUS    VERSION   AGE
istio-tenant-a   1           1       0        istio-tenant-a    Healthy   v1.24.1   30s
----
+
[NOTE]
====
The naming format of your revisions depends on which upgrade strategy you choose for your `Istio` instance.
====

. Copy the `ACTIVE REVISION` to use as your `istio.io/rev` label in the next step.

. Update injection labels on the `dataplane` namespace by running the following command:
+
[source,terminal]
----
$ oc label ns bookinfo istio.io/rev=istio-tenant-a maistra.io/ignore-namespace="true" --overwrite=true
----
+
This adds the following labels to the namespace:
+

> .. The `istio.io/rev: istio-tenant-a` label: Ensures that any new pods that get created in that namespace connect to the {SMProduct} 3.0 proxy.
> +
> .. The `maistra.io/ignore-namespace: "true"` label: Disables sidecar injection for {SMProduct} 2.6 proxies in the namespace so {SMProduct} 2.6 stops injecting proxies in this namespace, and any new proxies are injected by {SMProduct} 3.0. Without this label, the {SMProduct} 2.6 injection webhook tries to inject the pod and the injected sidecar proxy will refuse to start since it will have both the {SMProduct} 2.6 and the {SMProduct} 3.0 Container Network Interface(CNI) annotations.

Reviewer 3: ```suggestion
.. The `istio.io/rev: istio-tenant-a` label: Ensures that any new pods that get created in that namespace connect to the {SMProduct} 3.0 proxy.
+
.. The `maistra.io/ignore-namespace: "true"` label: Disables sidecar injection for {SMProduct} 2.6 proxies in the namespace so {SMProduct} 2.6 stops injecting proxies in this namespace, and any new proxies are injected by {SMProduct} 3.0. Without this label, the {SMProduct} 2.6 injection webhook tries to inject the pod and the injected sidecar proxy refuses to start since it has both the {SMProduct} 2.6 and the {SMProduct} 3.0 Container Network Interface(CNI) annotations.
```

+
[NOTE]
====

> Once you apply the `maistra.io/ignore-namespace` label, any new pod that gets created in the namespace will connect to the {SMProduct} 3.0 proxy. Workloads can still communicate with each other regardless of which controlplane they are connected to.

Reviewer 3: ```suggestion
Once you apply the `maistra.io/ignore-namespace` label, any new pod that gets created in the namespace connects to the {SMProduct} 3.0 proxy. Workloads can still communicate with each other regardless of which control plane they are connected to.
```

====

> . Restart the workloads:

Reviewer 3: This was what I intended in the previous module. Looks good here.

> . Restart the workloads:

Reviewer 3: ```suggestion
. Restart the workloads by using any one of the following options:
```

+

> .. You can restart all the workloads at once so that the new pods are injected with the {SMProduct} 3.0 proxy. For this example with the `bookinfo` application, run the following command:

Reviewer 3: ```suggestion
** To restart all the workloads at once so that the new pods are injected with the {SMProduct} 3.0 proxy, run the following command:
```

+

> [source,terminal]

Reviewer 3: ```suggestion
.Example command for `bookinfo` application
[source,terminal]
```

----
$ oc rollout restart deployments -n bookinfo
----

> .. Alternatively, you can restart each workload individually by running the following command for each workload:

Reviewer 3: ```suggestion
** To restart each workload individually, run the following command for each workload:
```

+

> [source,terminal]

Reviewer 3: ```suggestion
.Example command with `bookinfo` application
[source,terminal]
```

----
$ oc rollout restart deployments productpage-v1 -n bookinfo
----

. Wait for the `productpage` application to restart by running the following command:
+
[source,terminal]
----
$ oc rollout status deployment productpage-v1 -n bookinfo
----

> === Validate Workload Migration

Reviewer 3: Verification section seems to be a better option.

. Check that your workload is connected to the new control plane.

> .. See which proxies are still connected to the 2.6 control plane with the `istioctl` tool by running the following command:

Reviewer 3: ```suggestion
.. Fetch the list of proxies that are still connected to the 2.6 control plane with the `istioctl` tool by running the following command:
```

+

> [source,terminal]

Reviewer 3: ```suggestion
.Example command with the ServiceMeshControlPlane name set to basic
[source,terminal]
```

----
+
In this example, `basic` is the name of your `ServiceMeshControlPlane`:
+
.Example output
+
[source,terminal]
----
   NAME                                              CLUSTER        CDS        LDS        EDS        RDS          ECDS         ISTIOD                            VERSION
   details-v1-7b49464bc-zr7nr.bookinfo               Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   ratings-v1-d6f449f59-9rds2.bookinfo               Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v1-686cd989df-9x59z.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v2-785b8b48fc-l7xkj.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
   reviews-v3-67889ffd49-7bhxn.bookinfo              Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED       NOT SENT     istiod-basic-6c9f8d9894-sh6lx     1.20.8
----

> .. Check which proxies have been migrated to the new {SMProduct} 3.0 control plane by running the following command:

Reviewer 3: ```suggestion
.. View the list proxies that have been migrated to the new {SMProduct} 3.0 control plane by running the following command:
```
